test/metabase/middleware_test.clj

```diff
@@ -247,36 +247,35 @@
 ;; test that handler is killed when connection closes
 ;; Note: this closing over state is a dirty hack and the whole test should be
 ;; rewritten.
-(def test-slow-handler-state1 (atom :unset))
-(def test-slow-handler-state2 (atom :unset))
 
 (defn- test-slow-handler [state]
   (fn [_]
     (log/debug (u/format-color 'yellow "starting test-slow-handler"))
+    (reset! state :test-started)
     (Thread/sleep 7000)  ;; this is somewhat long to make sure the keepalive polling has time to kill it.
     (reset! state :ran-to-compleation)
     (log/debug (u/format-color 'yellow "finished test-slow-handler"))
     (resp/response {:success true})))
 
 (defn- start-and-maybe-kill-test-request [state kill?]
-  (reset! state [:initial-state kill?])
+  (reset! state :initial-state)
   (let [path "test-slow-handler"]
     (with-redefs [metabase.routes/routes (compojure.core/routes
                                           (GET (str "/" path) [] (middleware/streaming-json-response
                                                                   (test-slow-handler state))))]
       (let  [reader (io/input-stream (str "http://localhost:" (config/config-int :mb-jetty-port) "/" path))]
-        (Thread/sleep 1500)
+        (while (= :initial-state @state))
```

Comment:
Might want to have some sort of escape hatch here, otherwise we might just get a hang and not know why

Reply:
Any ideas what? If it's time based it feels like we are where we started.

Reply:
I usually would put something ridiculous in there, like 10 minutes. The reason I do that kind of thing is if there is some exception or something that prevents `@state` from ever equaling :intial-state in ci, you'll never know, it'll just time the build out eventually, but you won't know why. Locally you could maybe use `jstack` or something to figure out what the threads are doing and go from there.

Reply:
I was thinking about the problem wrong. You of course don't need sleep machinations, just timekeeping. Added. Thanks.